lib/web/apps.go

```diff
@@ -161,31 +163,41 @@ func (h *Handler) getAppDetails(w http.ResponseWriter, r *http.Request, p httpro
 		return nil, trace.Wrap(err, "unable to resolve FQDN: %v", req.FQDNHint)
 	}
 
+	fqdn := result.FQDN
+
+	proxyPublicAddr := utils.InferProxyPublicAddr(req.FQDNHint, h.proxyDNSNames())
+	alwaysUseProxyPublicAddr := result.App.GetAlwaysUseProxyPublicAddr()
+	if alwaysUseProxyPublicAddr {
+		appName := strings.TrimSuffix(req.FQDNHint, proxyPublicAddr)
+		fqdn = utils.AssembleAppFQDN(appName, proxyPublicAddr, clusterName, result.App)
+	}
+
 	resp := &GetAppDetailsResponse{
-		FQDN: result.FQDN,
+		FQDN: fqdn,
 	}
 
 	requiredAppNames := result.App.GetRequiredAppNames()
 
 	if !isRedirectFlow {
-		// TODO (avatus) this would be nice if the string in the RequiredApps spec was the fqdn of the required app
-		// so we could skip the resolution step all together but this would break existing configs.
-
-		// if clusterName is not supplied in the params, the initial app must have been fetched with fqdn hint only.
-		// We can use the clusterName of the initially resolved app
-		if clusterName == "" {
-			clusterName = result.ClusterName
-		}
-		for _, required := range requiredAppNames {
-			res, err := h.resolveApp(r.Context(), ctx, ResolveAppParams{ClusterName: clusterName, AppName: required})
-			if err != nil {
-				h.logger.ErrorContext(r.Context(), "Error getting app details for associated required app", "required_app", required, "app", result.App.GetName())
-				continue
+		for _, requiredAppName := range requiredAppNames {
+			// craft the fqdn manually so we do not have to resolve required apps
+			if alwaysUseProxyPublicAddr {
+				requiredAppFQDN := fmt.Sprintf("%s.%s", requiredAppName, proxyPublicAddr)
+				resp.RequiredAppFQDNs = append(resp.RequiredAppFQDNs, requiredAppFQDN)
+			} else {
+				if clusterName == "" {
+					clusterName = result.ClusterName
+				}
+				res, err := h.resolveApp(r.Context(), ctx, ResolveAppParams{ClusterName: clusterName, AppName: requiredAppName})
+				if err != nil {
+					h.logger.ErrorContext(r.Context(), "Error getting app details for associated required app", "required_app", requiredAppName, "app", result.App.GetName())
```

Comment:
```suggestion
					h.logger.ErrorContext(r.Context(), "Error getting app details for associated required app", "required_app", requiredAppName, "app", result.App.GetName(), "error", err)
```
Would logging error stack helpful to debug?